File: avalon-modules/avalon-dataset/avalon-dataset-controller/src/test/java/io/github/lishangbu/avalon/dataset/controller/TypeControllerTest.java
package io.github.lishangbu.avalon.dataset.controller;

import static org.junit.jupiter.api.Assertions.assertSame;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.github.lishangbu.avalon.dataset.entity.Type;
import io.github.lishangbu.avalon.dataset.service.TypeService;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

@ExtendWith(MockitoExtension.class)
class TypeControllerTest {
    @Mock private TypeService typeService;
    @InjectMocks private TypeController typeController;

    @Test
    void getTypePage_delegatesToService() {
        Type type = new Type();
        Pageable pageable = PageRequest.of(0, 5);
        Page<Type> page = new PageImpl<>(List.of(type));
        when(typeService.getPageByCondition(any(Type.class), eq(pageable))).thenReturn(page);

        Page<Type> result = typeController.getTypePage(pageable, type);

        assertSame(page, result);
        verify(typeService).getPageByCondition(type, pageable);
    }

    @Test
    void listTypes_delegatesToService() {
        Type type = new Type();
        List<Type> list = List.of(type);
        when(typeService.listByCondition(type)).thenReturn(list);

        List<Type> result = typeController.listTypes(type);

        assertSame(list, result);
        verify(typeService).listByCondition(type);
    }

    @Test
    void save_delegatesToService() {
        Type type = new Type();
        when(typeService.save(type)).thenReturn(type);

        Type result = typeController.save(type);

        assertSame(type, result);
        verify(typeService).save(type);
    }

    @Test
    void update_delegatesToService() {
        Type type = new Type();
        when(typeService.update(type)).thenReturn(type);

        Type result = typeController.update(type);

        assertSame(type, result);
        verify(typeService).update(type);
    }

    @Test
    void deleteById_delegatesToService() {
        typeController.deleteById(1L);
        verify(typeService).removeById(1L);
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-controller/src/test/kotlin/io/github/lishangbu/avalon/dataset/controller/TypeControllerTest.kt
package io.github.lishangbu.avalon.dataset.controller

import io.github.lishangbu.avalon.dataset.entity.dto.SaveTypeInput
import io.github.lishangbu.avalon.dataset.entity.dto.TypeSpecification
import io.github.lishangbu.avalon.dataset.entity.dto.TypeView
import io.github.lishangbu.avalon.dataset.entity.dto.UpdateTypeInput
import io.github.lishangbu.avalon.dataset.service.TypeService
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Test

class TypeControllerTest {
    @Test
    fun listTypes_delegatesToService() {
        val service = FakeTypeService()
        val controller = TypeController(service)
        val list = listOf(TypeView("1", "fire", "火"))
        service.listResult = list
        val specification = TypeSpecification(id = "1", internalName = "fire", name = "火")

        val result = controller.listTypes(specification)

        assertSame(list, result)
        assertEquals("1", service.listCondition!!.id)
    }

    @Test
    fun save_delegatesToService() {
        val service = FakeTypeService()
        val controller = TypeController(service)
        val command = SaveTypeInput("fire", "火")
        service.saveResult = TypeView("1", "fire", "火")

        val result = controller.save(command)

        assertSame(service.saveResult, result)
        assertSame(command, service.savedCommand)
    }

    @Test
    fun update_delegatesToService() {
        val service = FakeTypeService()
        val controller = TypeController(service)
        val command = UpdateTypeInput("1", "fire", "火")
        service.updateResult = TypeView("1", "fire", "火")

        val result = controller.update(command)

        assertSame(service.updateResult, result)
        assertSame(command, service.updatedCommand)
    }

    @Test
    fun deleteById_delegatesToService() {
        val service = FakeTypeService()
        val controller = TypeController(service)

        controller.deleteById(1L)

        assertEquals(1L, service.removedId)
    }

    private class FakeTypeService : TypeService {
        var listCondition: TypeSpecification? = null
        var savedCommand: SaveTypeInput? = null
        var updatedCommand: UpdateTypeInput? = null
        var removedId: Long? = null

        var listResult: List<TypeView> = emptyList()
        lateinit var saveResult: TypeView
        lateinit var updateResult: TypeView

        override fun save(command: SaveTypeInput): TypeView {
            savedCommand = command
            return saveResult
        }

        override fun update(command: UpdateTypeInput): TypeView {
            updatedCommand = command
            return updateResult
        }

        override fun removeById(id: Long) {
            removedId = id
        }

        override fun listByCondition(specification: TypeSpecification): List<TypeView> {
            listCondition = specification
            return listResult
        }
    }
}
